### PR TITLE
KTOR-372 Fix routing with trailing slash in path

### PR DESCRIPTION
**Pull request by rsinukov**
**Subsystem**
Server

**Motivation**
[Ktor should resolve trailing slash as a different URL
](https://youtrack.jetbrains.com/issue/KTOR-372)

**Solution**
Save flag if route ends with trailing slash and check it in selectors

### CODE REVIEW

File: ktor-server/ktor-server-core/api/ktor-server-core.api [hunk 1]
```diff
@@ -1264,13 +1264,13 @@ public final class io/ktor/routing/PathSegmentConstantRouteSelector : io/ktor/ro
 }
 
 public final class io/ktor/routing/PathSegmentOptionalParameterRouteSelector : io/ktor/routing/RouteSelector {
-	public fun <init> (Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;)V
-	public synthetic fun <init> (Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;ILkotlin/jvm/internal/DefaultConstructorMarker;)V
+	public fun <init> (Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Z)V
```
Comment (e5l) [inline ktor-server/ktor-server-core/api/ktor-server-core.api:1267]: These breaking changes should be fixed to fit in `1.5.0`: no removes required

File: ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RouteSelector.kt [hunk 2]
```diff
@@ -119,6 +120,9 @@ class RootRouteSelector(rootPath: String = "") : RouteSelector(RouteSelectorEval
         if (segments.size < parts.size) {
             return RouteSelectorEvaluation.Failed
         }
+        if (hasTrailingSlash != context.hasTrailingSlash) {
```
Comment (e5l) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RouteSelector.kt:123]: `||`

File: ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RouteSelector.kt [hunk 3]
```diff
@@ -181,11 +185,19 @@ data class OptionalParameterRouteSelector(val name: String) : RouteSelector(Rout
...
+data class PathSegmentConstantRouteSelector(
+    val value: String,
+    private val hasTrailingSlash: Boolean
+) : RouteSelector(RouteSelectorEvaluation.qualityConstant) {
+
     override fun evaluate(context: RoutingResolveContext, segmentIndex: Int): RouteSelectorEvaluation {
-        if (segmentIndex < context.segments.size && context.segments[segmentIndex] == value)
-            return RouteSelectorEvaluation.ConstantPath
-        return RouteSelectorEvaluation.Failed
+        return when {
```
Comment (e5l) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RouteSelector.kt:194]: can be an expression, but I don't insist

File: ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RoutingPath.kt [hunk 4]
```diff
@@ -11,18 +11,18 @@ import io.ktor.http.*
  *
  * @property parts contains parsed routing path segments
  */
-class RoutingPath private constructor(val parts: List<RoutingPathSegment>) {
-    companion object {
+public class RoutingPath private constructor(val parts: List<RoutingPathSegment>) {
```
Comment (e5l) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RoutingPath.kt:14]: Should be the separate changeset as discussed

**Review by e5l — CHANGES_REQUESTED** [review]

File: ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RouteSelector.kt [hunk 5]
```diff
@@ -259,6 +329,28 @@ data class PathSegmentOptionalParameterRouteSelector(val name: String, val prefi
...
+        if (name != other.name) return false
+        if (prefix != other.prefix) return false
+        if (suffix != other.suffix) return false
+        if (hasTrailingSlash != other.hasTrailingSlash) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = name.hashCode()
```
Comment (e5l) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RouteSelector.kt:348]: consider extracting `combine` for hashcodes

File: ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RouteSelector.kt [hunk 6]
```diff
@@ -259,6 +329,28 @@ data class PathSegmentOptionalParameterRouteSelector(val name: String, val prefi
     }
 
     override fun toString(): String = "${prefix ?: ""}{$name?}${suffix ?: ""}"
+
+    override fun equals(other: Any?): Boolean {
```
Comment (e5l) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RouteSelector.kt:333]: I guess all equals and hashcodes should have tests

**Review by e5l — CHANGES_REQUESTED** [review]

File: ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RouteSelector.kt (hunk 5 above)
Comment (rsinukov) [reply ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RouteSelector.kt:354]: this code is autogenerated by IDE and I would rather avoid making changes there. As discussed offline, I added comment about this.

File: ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RouteSelector.kt (hunk 6 above)
Comment (rsinukov) [reply ktor-server/ktor-server-core/jvm/src/io/ktor/routing/RouteSelector.kt:338]: same here

**Review by e5l — APPROVED** [review]
